ICE on beta and nightly as result of use of `yield` in combination with macro call in `async` block

This code 
```rust
fn main() {
    async { yield print!(":C") };
}

```
([Playground](https://play.rust-lang.org/?version=beta&mode=debug&edition=2018&gist=7162a06063ca516e9f2f4399f6ffffb4))

produces ICE on current beta and nightly 
```
thread 'rustc' panicked at 'src/librustc/hir/map/hir_id_validator.rs:26: 
ItemLocalIds not assigned densely in ::main[0]. Max ItemLocalId = 14, missing IDs = ["[local_id: 1, node:unknown node (hir_id=HirId { owner: DefIndex(12), local_id: 1 })]"]; seens IDs = ["(HirId { owner: DefIndex(12), local_id: 0 } fn main (hir_id=HirId { owner: DefIndex(12), local_id: 0 }))", "(HirId { owner: DefIndex(12), local_id: 13 } block { ::std::future::from_generator(|| { (/*ERROR*/) }); } (hir_id=HirId { owner: DefIndex(12), local_id: 13 }))", "(HirId { owner: DefIndex(12), local_id: 10 } expr ::std::future::from_generator (hir_id=HirId { owner: DefIndex(12), local_id: 10 }))", "(HirId { owner: DefIndex(12), local_id: 7 } path segment std (hir_id=HirId { owner: DefIndex(12), local_id: 7 }))", "(HirId { owner: DefIndex(12), local_id: 4 } expr { (/*ERROR*/) } (hir_id=HirId { owner: DefIndex(12), local_id: 4 }))", "(HirId { owner: DefIndex(12), local_id: 14 } expr { ::std::future::from_generator(|| { (/*ERROR*/) }); } (hir_id=HirId { owner: DefIndex(12), local_id: 14 }))", "(HirId { owner: DefIndex(12), local_id: 11 } expr ::std::future::from_generator(|| { (/*ERROR*/) }) (hir_id=HirId { owner: DefIndex(12), local_id: 11 }))", "(HirId { owner: DefIndex(12), local_id: 8 } path segment future (hir_id=HirId { owner: DefIndex(12), local_id: 8 }))", "(HirId { owner: DefIndex(12), local_id: 5 } expr || { (/*ERROR*/) } (hir_id=HirId { owner: DefIndex(12), local_id: 5 }))", "(HirId { owner: DefIndex(12), local_id: 2 } expr (/*ERROR*/) (hir_id=HirId { owner: DefIndex(12), local_id: 2 }))", "(HirId { owner: DefIndex(12), local_id: 12 } stmt ::std::future::from_generator(|| { (/*ERROR*/) }); (hir_id=HirId { owner: DefIndex(12), local_id: 12 }))", "(HirId { owner: DefIndex(12), local_id: 9 } path segment from_generator (hir_id=HirId { owner: DefIndex(12), local_id: 9 }))", "(HirId { owner: DefIndex(12), local_id: 6 } path segment  (hir_id=HirId { owner: DefIndex(12), local_id: 6 }))", "(HirId { owner: DefIndex(12), local_id: 3 } block { (/*ERROR*/) } (hir_id=HirId { owner: DefIndex(12), local_id: 3 }))"]', src/librustc/util/bug.rs:37:26
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.40.0-beta.2 (2a0ba54fa 2019-11-16) running on x86_64-apple-darwin

note: compiler flags: -C debuginfo=2 -C incremental --crate-type bin

note: some of the compiler flags provided by cargo are hidden

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0658`.
error: could not compile `rust-games`.
```

## Regression found in the compiler

searched nightlies: from nightly-2019-01-01 to nightly-2020-02-01
regressed nightly: nightly-2019-09-29
searched commits: from https://github.com/rust-lang/rust/commit/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c to https://github.com/rust-lang/rust/commit/488381ce9ef0ceabe83b73127c659e5d38137df0
regressed commit: https://github.com/rust-lang/rust/commit/f3c8eba643a815d720e7f20699b3dca144c845c4
source code: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=956b959a5649adaecb4fe39acef8239d

## Instructions

```
cargo bisect-rustc --start 2019-09-28 --end 2019-09-29 --prompt
```


## Error

<details><summary>COLLAPSIBLE ERROR STACKTRACE</summary>
<p>

```bash
error[E0658]: yield syntax is experimental
 --> src/main.rs:2:13
  |
2 |     async { yield print!(":C") };
  |             ^^^^^^^^^^^^^^^^^^
  |
  = note: for more information, see https://github.com/rust-lang/rust/issues/43122
  = help: add `#![feature(generators)]` to the crate attributes to enable

error[E0727]: `async` generators are not yet supported
 --> src/main.rs:2:13
  |
2 |     async { yield print!(":C") };
  |             ^^^^^^^^^^^^^^^^^^

thread 'rustc' panicked at 'src/librustc/hir/map/hir_id_validator.rs:27: 
ItemLocalIds not assigned densely in ::main[0]. Max ItemLocalId = 14, missing IDs = ["[local_id: 1, node:unknown node (hir_id=HirId { owner: DefIndex(3), local_id: 1 })]"]; seens IDs = ["(HirId { owner: DefIndex(3), local_id: 0 } fn main (hir_id=HirId { owner: DefIndex(3), local_id: 0 }))", "(HirId { owner: DefIndex(3), local_id: 13 } block { ::std::future::from_generator(|| { (/*ERROR*/) }); } (hir_id=HirId { owner: DefIndex(3), local_id: 13 }))", "(HirId { owner: DefIndex(3), local_id: 10 } expr ::std::future::from_generator (hir_id=HirId { owner: DefIndex(3), local_id: 10 }))", "(HirId { owner: DefIndex(3), local_id: 7 } path segment std (hir_id=HirId { owner: DefIndex(3), local_id: 7 }))", "(HirId { owner: DefIndex(3), local_id: 4 } expr { (/*ERROR*/) } (hir_id=HirId { owner: DefIndex(3), local_id: 4 }))", "(HirId { owner: DefIndex(3), local_id: 14 } expr { ::std::future::from_generator(|| { (/*ERROR*/) }); } (hir_id=HirId { owner: DefIndex(3), local_id: 14 }))", "(HirId { owner: DefIndex(3), local_id: 11 } expr ::std::future::from_generator(|| { (/*ERROR*/) }) (hir_id=HirId { owner: DefIndex(3), local_id: 11 }))", "(HirId { owner: DefIndex(3), local_id: 8 } path segment future (hir_id=HirId { owner: DefIndex(3), local_id: 8 }))", "(HirId { owner: DefIndex(3), local_id: 5 } expr || { (/*ERROR*/) } (hir_id=HirId { owner: DefIndex(3), local_id: 5 }))", "(HirId { owner: DefIndex(3), local_id: 2 } expr (/*ERROR*/) (hir_id=HirId { owner: DefIndex(3), local_id: 2 }))", "(HirId { owner: DefIndex(3), local_id: 12 } stmt ::std::future::from_generator(|| { (/*ERROR*/) }); (hir_id=HirId { owner: DefIndex(3), local_id: 12 }))", "(HirId { owner: DefIndex(3), local_id: 9 } path segment from_generator (hir_id=HirId { owner: DefIndex(3), local_id: 9 }))", "(HirId { owner: DefIndex(3), local_id: 6 } path segment  (hir_id=HirId { owner: DefIndex(3), local_id: 6 }))", "(HirId { owner: DefIndex(3), local_id: 3 } block { (/*ERROR*/) } (hir_id=HirId { owner: DefIndex(3), local_id: 3 }))"]', src/librustc/util/bug.rs:37:26
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:77
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1052
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1428
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:204
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:224
  10: rustc_driver::report_ice
  11: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:474
  12: std::panicking::begin_panic
  13: rustc::util::bug::opt_span_bug_fmt::{{closure}}
  14: rustc::ty::context::tls::with_opt::{{closure}}
  15: rustc::ty::context::tls::with_opt
  16: rustc::util::bug::opt_span_bug_fmt
  17: rustc::util::bug::bug_fmt
  18: rustc::hir::map::hir_id_validator::check_crate
  19: rustc_session::utils::<impl rustc_session::session::Session>::time
  20: rustc::hir::map::map_crate
  21: rustc_interface::passes::create_global_ctxt
  22: rustc_interface::queries::Queries::global_ctxt
  23: rustc_interface::interface::run_compiler_in_existing_thread_pool
  24: scoped_tls::ScopedKey<T>::set
  25: syntax::attr::with_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.43.0-nightly (58b834344 2020-02-05) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental --crate-type bin

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
error: aborting due to 2 previous errors

Some errors have detailed explanations: E0658, E0727.
For more information about an error, try `rustc --explain E0658`.

```

</p></details>

--- a/src/librustc/hir/map/hir_id_validator.rs
+++ b/src/librustc/hir/map/hir_id_validator.rs
@@ -7,7 +7,7 @@ use rustc_hir::intravisit;
 use rustc_hir::itemlikevisit::ItemLikeVisitor;
 use rustc_hir::{HirId, ItemLocalId};
 
-pub fn check_crate(hir_map: &Map<'_>) {
+pub fn check_crate(hir_map: &Map<'_>, sess: &rustc_session::Session) {
     hir_map.dep_graph.assert_ignored();
 
     let errors = Lock::new(Vec::new());
@@ -24,7 +24,7 @@ pub fn check_crate(hir_map: &Map<'_>) {
 
     if !errors.is_empty() {
         let message = errors.iter().fold(String::new(), |s1, s2| s1 + "\n" + s2);
-        bug!("{}", message);
+        sess.delay_span_bug(rustc_span::DUMMY_SP, &message);
     }
 }
 

--- a/src/librustc/hir/map/mod.rs
+++ b/src/librustc/hir/map/mod.rs
@@ -1235,7 +1235,7 @@ pub fn map_crate<'hir>(
     let map = Map { krate, dep_graph, crate_hash, map, hir_to_node_id, definitions };
 
     sess.time("validate_HIR_map", || {
-        hir_id_validator::check_crate(&map);
+        hir_id_validator::check_crate(&map, sess);
     });
 
     map

--- a/src/test/ui/generator/async-generator-issue-67158.rs
+++ b/src/test/ui/generator/async-generator-issue-67158.rs
@@ -1,0 +1,6 @@
+#![feature(generators)]
+// edition:2018
+// Regression test for #67158.
+fn main() {
+    async { yield print!(":C") }; //~ ERROR `async` generators are not yet supported
+}

--- a/src/test/ui/generator/async-generator-issue-67158.stderr
+++ b/src/test/ui/generator/async-generator-issue-67158.stderr
@@ -1,0 +1,9 @@
+error[E0727]: `async` generators are not yet supported
+  --> $DIR/async-generator-issue-67158.rs:5:13
+   |
+LL |     async { yield print!(":C") };
+   |             ^^^^^^^^^^^^^^^^^^
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0727`.

